File: java-workshop/part-4-behavior/src/test/java/service/OrderServiceTest.java
package service;

import domain.Order;
import domain.OrderItem;
import domain.OrderStatus;
import helper.InMemoryOrderRepository;
import helper.InMemoryPaymentGateway;
import helper.InMemoryShippingGateway;
import helper.InMemoryStockRepository;
import helper.OrderBuilder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.fail;

class OrderServiceTest {

    private InMemoryOrderRepository orderRepository;
    private InMemoryPaymentGateway paymentGateway;
    private InMemoryShippingGateway shippingGateway;
    private InMemoryStockRepository stockRepository;
    private OrderService orderService;

    @BeforeEach
    void setUp() {
        orderRepository = new InMemoryOrderRepository();
        stockRepository = new InMemoryStockRepository();
        paymentGateway = new InMemoryPaymentGateway();
        shippingGateway = new InMemoryShippingGateway();

        orderService = new OrderService(
            orderRepository,
            new RepositoryStockService(stockRepository),
            new PricingService(),
            new GatewayPaymentService(paymentGateway),
            new GatewayShippingService(shippingGateway)
        );
    }

    // ══════════════════════════════════════════════════════════════════════════
    // YOUR EXERCISE — See README for full instructions
    // ══════════════════════════════════════════════════════════════════════════

    // --- Step 1: Reject already-used discount codes ---
    // 1.4. Create an InMemoryDiscountCodeRepository (already provided in helpers).
    //      Wire a DiscountCodeService into OrderService via the constructor.
    //      Create a given helper to set up the discount code state.
    // 1.5. Fill in the GIVEN/WHEN/THEN below

    @Disabled
    @Test
    void p4_orderIsRejected_whenDiscountCodeIsAlreadyUsed() {
        // GIVEN an order with discount code SUMMER20
        // AND this code has already been used by this customer

        // WHEN the customer places the order

        // THEN the order is rejected with reason "Discount code already used"
        // AND no payment is triggered

        fail("TODO: implement scenario 'discount code already used'");
    }
    // ✅ 1.6. Run: ./gradlew test — then back to README for Step 2

    // --- Step 2: Apply discount to the price ---
    // 2.3. Create a given helper for a valid discount code.
    //      Don't forget to set up sufficient stock.
    // 2.4. Fill in the GIVEN/WHEN/THEN below

    @Disabled
    @Test
    void p4_orderIsConfirmed_whenDiscountCodeIsValid() {
        // GIVEN an order with 2 items at €55 each (subtotal: €110)
        // AND discount code SUMMER20 (-20%)
        // AND the code has not yet been used by this customer
        // AND sufficient stock

        // WHEN the customer places the order

        // THEN payment is processed for €105.60
        // AND the order is confirmed

        fail("TODO: implement scenario 'order with a valid discount code'");
    }

    // ✅ 2.5. Run: ./gradlew test — then back to README for Step 3

    // --- Step 3: Mark as used after payment ---
    // 3.4. Update the test above to also verify that the discount code is marked as used.
    //      Hint: assertTrue(discountCodeRepository.isUsed("CUST-001", DiscountCode.SUMMER20));
    // 3.5. Fill in the GIVEN/WHEN/THEN below

    @Disabled
    @Test
    void p4_discountCodeIsMarkedAsUsed_afterSuccessfulPayment() {
        // GIVEN an order with a valid discount code
        // AND payment succeeds

        // WHEN the customer places the order

        // THEN the discount code is marked as used

        fail("TODO: implement scenario 'discount code marked as used after payment'");
    }
    // ✅ 3.6. Run: ./gradlew test ✅

    // ══════════════════════════════════════════════════════════════════════════
    // REFERENCE TESTS — existing tests and helpers, for inspiration
    // ══════════════════════════════════════════════════════════════════════════

    // ── Given helpers ─────────────────────────────────────────────────────────

    private OrderBuilder anOrder() {
        return OrderBuilder.anOrder()
            .withId("ORDER-001")
            .withCustomerId("CUST-001")
            .withItems(List.of(new OrderItem("PROD-001", 1, 20.0)));
    }

    private void givenStockIsAvailableFor(Order order) {
        order.items().forEach(item -> stockRepository.withAvailableStock(item.productId(), item.quantity()));
    }

    private void givenPaymentIsDeclined() {
        paymentGateway.willDecline();
    }

    // ── Validation ────────────────────────────────────────────────────────────

    @Test
    void p4_orderIsRejected_whenOrderIsAlreadyConfirmed() {
        var order = anOrder().withStatus(OrderStatus.CONFIRMED).build();

        OrderResult result = orderService.placeOrder(order);

        assertInstanceOf(OrderResult.Failure.class, result);
        assertEquals("Invalid order", ((OrderResult.Failure) result).reason());
    }

    @Test
    void p4_orderIsRejected_whenOrderHasNoItems() {
        var order = anOrder().withItems(List.of()).build();

        OrderResult result = orderService.placeOrder(order);

        assertInstanceOf(OrderResult.Failure.class, result);
        assertEquals("Invalid order", ((OrderResult.Failure) result).reason());
    }

    @Test
    void p4_noPaymentIsCharged_whenOrderIsInvalid() {
        var order = anOrder().withItems(List.of()).build();

        orderService.placeOrder(order);

        assertFalse(paymentGateway.wasCharged());
    }

    // ── Happy path ────────────────────────────────────────────────────────────

    @Test
    void p4_orderIsConfirmed_whenAllStepsSucceed() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);

        OrderResult result = orderService.placeOrder(order);

        assertInstanceOf(OrderResult.Success.class, result);
        assertEquals(OrderStatus.CONFIRMED, ((OrderResult.Success) result).order().status());
    }

    @Test
    void p4_calculatedPriceIsStoredInOrder_whenAllStepsSucceed() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);

        OrderResult result = orderService.placeOrder(order);

        double realTotal = 29.0; // subtotal=20, tax=4 (20%), shipping=5 (below 100€ threshold)
        assertEquals(realTotal, ((OrderResult.Success) result).order().totalPrice());
    }

    @Test
    void p4_shippingConfirmationIsReturned_whenAllStepsSucceed() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);

        OrderResult result = orderService.placeOrder(order);

        assertEquals(shippingGateway.lastConfirmation(), ((OrderResult.Success) result).shippingConfirmation());
    }

    // ── Stock unavailable ─────────────────────────────────────────────────────

    @Test
    void p4_orderFails_whenStockIsUnavailable() {
        var order = anOrder().build();
        // stockRepository is empty: no stock added

        OrderResult result = orderService.placeOrder(order);

        assertInstanceOf(OrderResult.Failure.class, result);
        assertEquals("Insufficient stock", ((OrderResult.Failure) result).reason());
    }

    @Test
    void p4_paymentIsNotCharged_whenStockIsUnavailable() {
        var order = anOrder().build();

        orderService.placeOrder(order);

        assertFalse(paymentGateway.wasCharged());
    }

    @Test
    void p4_shipmentIsNotCreated_whenStockIsUnavailable() {
        var order = anOrder().build();

        orderService.placeOrder(order);

        assertFalse(shippingGateway.hasCreatedShipment());
    }

    // ── Payment declined ──────────────────────────────────────────────────────

    @Test
    void p4_orderFails_whenPaymentIsDeclined() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);
        givenPaymentIsDeclined();

        OrderResult result = orderService.placeOrder(order);

        assertInstanceOf(OrderResult.Failure.class, result);
        assertEquals("Payment failed", ((OrderResult.Failure) result).reason());
    }

    @Test
    void p4_shipmentIsNotCreated_whenPaymentIsDeclined() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);
        givenPaymentIsDeclined();

        orderService.placeOrder(order);

        assertFalse(shippingGateway.hasCreatedShipment());
    }

    @Test
    void p4_stockIsReleased_whenPaymentIsDeclined() {
        var order = anOrder().build();
        givenStockIsAvailableFor(order);
        givenPaymentIsDeclined();

        orderService.placeOrder(order);

        order.items().forEach(item ->
            assertEquals(item.quantity(), stockRepository.availableStock(item.productId())));
    }
}

File: csharp-workshop/Part4.Behavior.Tests/Service/OrderServiceTest.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Service;
using Part4.Behavior.Tests.Helper;
using Xunit;

namespace Part4.Behavior.Tests.Service;

public class OrderServiceTest
{
    private readonly InMemoryOrderRepository _orderRepository;
    private readonly InMemoryPaymentGateway _paymentGateway;
    private readonly InMemoryShippingGateway _shippingGateway;
    private readonly InMemoryStockRepository _stockRepository;
    private readonly OrderService _orderService;

    public OrderServiceTest()
    {
        _orderRepository = new InMemoryOrderRepository();
        _stockRepository = new InMemoryStockRepository();
        _paymentGateway = new InMemoryPaymentGateway();
        _shippingGateway = new InMemoryShippingGateway();

        _orderService = new OrderService(
            _orderRepository,
            new RepositoryStockService(_stockRepository),
            new PricingService(),
            new GatewayPaymentService(_paymentGateway),
            new GatewayShippingService(_shippingGateway)
        );
    }

    // ══════════════════════════════════════════════════════════════════════════
    // YOUR EXERCISE — See README for full instructions
    // ══════════════════════════════════════════════════════════════════════════

    // --- Step 1: Reject already-used discount codes ---
    // 1.3. Create an InMemoryDiscountCodeRepository (already provided in Helper/).
    //      Wire a DiscountCodeService into OrderService via the constructor.
    //      Create a given helper to set up the discount code state.
    // 1.4. Fill in the GIVEN/WHEN/THEN below
    [Fact(Skip = "TODO")]
    public void P4_OrderIsRejected_WhenDiscountCodeIsAlreadyUsed()
    {
        // GIVEN an order with discount code SUMMER20
        // AND this code has already been used by this customer

        // WHEN the customer places the order

        // THEN the order is rejected with reason "Discount code already used"
        // AND no payment is triggered
    }
    // ✅ 1.5. Run: dotnet test Part4.Behavior.Tests/ — then back to README for Step 2
    // ─── END OF STEP 1 ──────────────────────────────────────────────────────



    // --- Step 2: Apply discount to the price ---
    // 2.2. Create a given helper for a valid discount code.
    //      Don't forget to set up sufficient stock.
    // 2.3. Fill in the GIVEN/WHEN/THEN below
    [Fact(Skip = "TODO")]
    public void P4_OrderIsConfirmed_WhenDiscountCodeIsValid()
    {
        // GIVEN an order with 2 items at €55 each (subtotal: €110)
        // AND discount code SUMMER20 (-20%)
        // AND the code has not yet been used by this customer
        // AND sufficient stock

        // WHEN the customer places the order

        // THEN payment is processed for €105.60
        // AND the order is confirmed
    }

    // ✅ 2.4. Run: dotnet test Part4.Behavior.Tests/ — then back to README for Step 3
    // ─── END OF STEP 2 ──────────────────────────────────────────────────────



    // --- Step 3: Mark as used after payment ---
    // 3.2. Update the test above to also verify that the discount code is marked as used.
    //      Hint: Assert.True(_discountCodeRepository.IsUsed("CUST-001", DiscountCode.SUMMER20));
    // 3.3. Fill in the GIVEN/WHEN/THEN below
    [Fact(Skip = "TODO")]
    public void P4_DiscountCodeIsMarkedAsUsed_AfterSuccessfulPayment()
    {
        // GIVEN an order with a valid discount code
        // AND payment succeeds

        // WHEN the customer places the order

        // THEN the discount code is marked as used
    }
    // ✅ 3.4. Run: dotnet test Part4.Behavior.Tests/ ✅
    // ─── END OF STEP 3 ──────────────────────────────────────────────────────

    // ══════════════════════════════════════════════════════════════════════════
    // REFERENCE TESTS — existing tests and helpers, for inspiration
    // ══════════════════════════════════════════════════════════════════════════

    // ── Given helpers ─────────────────────────────────────────────────────────

    private static OrderBuilder AnOrder() =>
        OrderBuilder.AnOrder()
            .WithId("ORDER-001")
            .WithCustomerId("CUST-001")
            .WithItems(new List<OrderItem> { new("PROD-001", 1, 20.0) });

    private void GivenStockIsAvailableFor(Order order)
    {
        foreach (var item in order.Items)
        {
            _stockRepository.WithAvailableStock(item.ProductId, item.Quantity);
        }
    }

    private void GivenPaymentIsDeclined()
    {
        _paymentGateway.WillDecline();
    }

    // ── Validation ────────────────────────────────────────────────────────────

    [Fact]
    public void P4_OrderIsRejected_WhenOrderIsAlreadyConfirmed()
    {
        var order = AnOrder().WithStatus(OrderStatus.Confirmed).Build();

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.IsType<OrderResult.Failure>(result);
        Assert.Equal("Invalid order", ((OrderResult.Failure)result).Reason);
    }

    [Fact]
    public void P4_OrderIsRejected_WhenOrderHasNoItems()
    {
        var order = AnOrder().WithItems(new List<OrderItem>()).Build();

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.IsType<OrderResult.Failure>(result);
        Assert.Equal("Invalid order", ((OrderResult.Failure)result).Reason);
    }

    [Fact]
    public void P4_NoPaymentIsCharged_WhenOrderIsInvalid()
    {
        var order = AnOrder().WithItems(new List<OrderItem>()).Build();

        _orderService.PlaceOrder(order);

        Assert.False(_paymentGateway.WasCharged());
    }

    // ── Happy path ────────────────────────────────────────────────────────────

    [Fact]
    public void P4_OrderIsConfirmed_WhenAllStepsSucceed()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.IsType<OrderResult.Success>(result);
        Assert.Equal(OrderStatus.Confirmed, ((OrderResult.Success)result).Order.Status);
    }

    [Fact]
    public void P4_CalculatedPriceIsStoredInOrder_WhenAllStepsSucceed()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);

        OrderResult result = _orderService.PlaceOrder(order);

        double realTotal = 29.0; // subtotal=20, tax=4 (20%), shipping=5 (below 100€ threshold)
        Assert.Equal(realTotal, ((OrderResult.Success)result).Order.TotalPrice);
    }

    [Fact]
    public void P4_ShippingConfirmationIsReturned_WhenAllStepsSucceed()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.Equal(_shippingGateway.LastConfirmation(), ((OrderResult.Success)result).ShippingConfirmation);
    }

    // ── Stock unavailable ─────────────────────────────────────────────────────

    [Fact]
    public void P4_OrderFails_WhenStockIsUnavailable()
    {
        var order = AnOrder().Build();
        // stockRepository is empty: no stock added

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.IsType<OrderResult.Failure>(result);
        Assert.Equal("Insufficient stock", ((OrderResult.Failure)result).Reason);
    }

    [Fact]
    public void P4_PaymentIsNotCharged_WhenStockIsUnavailable()
    {
        var order = AnOrder().Build();

        _orderService.PlaceOrder(order);

        Assert.False(_paymentGateway.WasCharged());
    }

    [Fact]
    public void P4_ShipmentIsNotCreated_WhenStockIsUnavailable()
    {
        var order = AnOrder().Build();

        _orderService.PlaceOrder(order);

        Assert.False(_shippingGateway.HasCreatedShipment());
    }

    // ── Payment declined ──────────────────────────────────────────────────────

    [Fact]
    public void P4_OrderFails_WhenPaymentIsDeclined()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);
        GivenPaymentIsDeclined();

        OrderResult result = _orderService.PlaceOrder(order);

        Assert.IsType<OrderResult.Failure>(result);
        Assert.Equal("Payment failed", ((OrderResult.Failure)result).Reason);
    }

    [Fact]
    public void P4_ShipmentIsNotCreated_WhenPaymentIsDeclined()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);
        GivenPaymentIsDeclined();

        _orderService.PlaceOrder(order);

        Assert.False(_shippingGateway.HasCreatedShipment());
    }

    [Fact]
    public void P4_StockIsReleased_WhenPaymentIsDeclined()
    {
        var order = AnOrder().Build();
        GivenStockIsAvailableFor(order);
        GivenPaymentIsDeclined();

        _orderService.PlaceOrder(order);

        foreach (var item in order.Items)
        {
            Assert.Equal(item.Quantity, _stockRepository.AvailableStock(item.ProductId));
        }
    }
}
